### PR TITLE
fix: package hash bug

### PR DESCRIPTION
**Pull request by theoparis**
This is not the best fix atm as it duplicates code.
See #17064

### CODE REVIEW

File: src/Package.zig [hunk 1]
```diff
@@ -608,6 +608,20 @@ fn fetchAndUnpack(
...
+
+                const iterable_dir = try global_cache_directory.handle.makeOpenPathIterable(tmp_dir_sub_path, .{});
+                errdefer iterable_dir.close();
+
+                break :d .{
+                    .path = path,
+                    .handle = iterable_dir.dir,
+                };
+            };
+            defer tmp_directory.closeAndFree(gpa);
```
Comment (squeek502) [inline src/Package.zig:624]: Won't this double `closeAndFree` `tmp_directory`? And the first `tmp_directory` won't ever be `closeAndFree`d?

File: src/Package.zig [hunk 2]
```diff
@@ -608,6 +608,20 @@ fn fetchAndUnpack(
             // I have not checked what buffer sizes the xz decompression implementation uses
             // by default, so the same logic applies for buffering the reader as for gzip.
             try unpackTarball(gpa, prog_reader.reader(), tmp_directory.handle, std.compress.xz);
+
+            tmp_directory = d: {
```
Comment (squeek502) [inline src/Package.zig:612]: Why only reopen in the `application/x-xz` case? Wouldn't this affect all the other cases too?

Comment (squeek502) [inline src/Package.zig:612]: Definitely need an explainer comment detailing why this is needed (and that it's effectively a workaround for a potential IterableDir/Walker bug).

**Comment by andrewrk** [comment]
cleaner workaround coming as part of  #17392

**Pull request closed without merging**